File: backend/internal/sonos/device.go
package sonos

import (
	"context"
	"encoding/xml"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// HTTPPort is the unencrypted HTTP port every Sonos ZonePlayer exposes for
// SOAP control, the description XML, GENA event subscriptions and /status/*.
const HTTPPort = 1400

// Service describes a single UPnP service advertised by a ZonePlayer, taken
// from its device description XML. The control/event/SCPD URLs are paths
// relative to http://{ip}:1400.
type Service struct {
	Type       string // e.g. urn:schemas-upnp-org:service:AVTransport:1
	ID         string // e.g. urn:upnp-org:serviceId:AVTransport
	ControlURL string // e.g. /MediaRenderer/AVTransport/Control
	EventURL   string // e.g. /MediaRenderer/AVTransport/Event
	SCPDURL    string
}

// Device is a discovered Sonos ZonePlayer, populated from its
// device_description.xml. It is the per-physical-speaker handle; room/group
// layout comes separately from ZoneGroupTopology.
type Device struct {
	IP              net.IP
	UDN             string // uuid:RINCON_xxxxxxxxxxxx01400
	FriendlyName    string
	ModelName       string // e.g. "Sonos PLAY:1"
	ModelNumber     string // e.g. "S13"
	SoftwareVersion string // displayVersion, e.g. "15.9"
	RoomName        string // zone/room name if present in the description

	// Services indexed by short name (the trailing segment of the serviceId,
	// e.g. "AVTransport", "RenderingControl", "Queue").
	Services map[string]Service
}

// RINCON returns the bare RINCON UUID (without the "uuid:" prefix) used in
// x-rincon grouping URIs and topology coordinator references.
func (d *Device) RINCON() string {
	return strings.TrimPrefix(d.UDN, "uuid:")
}

// BaseURL is http://{ip}:1400 — the root every service path hangs off.
func (d *Device) BaseURL() string {
	return fmt.Sprintf("http://%s:%d", d.IP.String(), HTTPPort)
}

// Generation reports a best-effort S1/S2 guess. Sonos splits its ecosystem
// into S1 (legacy) and S2 (modern). The dedicated Queue service
// (urn:schemas-sonos-com:service:Queue:1) is present on S2 players, so we use
// it as the primary signal and fall back to the firmware major version. The
// returned string is one of "S2", "S1", or "unknown" — never silently
// defaulted; callers should show the raw SoftwareVersion alongside it.
func (d *Device) Generation() string {
	if _, ok := d.Services["Queue"]; ok {
		return "S2"
	}
	// Firmware major version: S1 caps out around v11.x; S2 is v12+.
	if maj, ok := firmwareMajor(d.SoftwareVersion); ok {
		if maj >= 12 {
			return "S2"
		}
		return "S1"
	}
	return "unknown"
}

func firmwareMajor(v string) (int, bool) {
	if v == "" {
		return 0, false
	}
	part, _, _ := strings.Cut(v, ".")
	n := 0
	for _, r := range part {
		if r < '0' || r > '9' {
			return 0, false
		}
		n = n*10 + int(r-'0')
	}
	return n, true
}

// --- description XML parsing ---

// xmlDevice mirrors the nested <device> structure in device_description.xml.
// Services are spread across the root device and embedded MediaRenderer /
// MediaServer devices, so we recurse.
type xmlDevice struct {
	DeviceType      string      `xml:"deviceType"`
	FriendlyName    string      `xml:"friendlyName"`
	ModelName       string      `xml:"modelName"`
	ModelNumber     string      `xml:"modelNumber"`
	DisplayVersion  string      `xml:"displayVersion"`
	SoftwareVersion string      `xml:"softwareVersion"`
	RoomName        string      `xml:"roomName"`
	UDN             string      `xml:"UDN"`
	ServiceList     []xmlSvc    `xml:"serviceList>service"`
	DeviceList      []xmlDevice `xml:"deviceList>device"`
}

type xmlSvc struct {
	ServiceType string `xml:"serviceType"`
	ServiceID   string `xml:"serviceId"`
	ControlURL  string `xml:"controlURL"`
	EventSubURL string `xml:"eventSubURL"`
	SCPDURL     string `xml:"SCPDURL"`
}

type xmlRoot struct {
	XMLName xml.Name  `xml:"root"`
	Device  xmlDevice `xml:"device"`
}

// shortServiceName turns a serviceId like
// "urn:upnp-org:serviceId:AVTransport" into "AVTransport".
func shortServiceName(serviceID string) string {
	if i := strings.LastIndexByte(serviceID, ':'); i >= 0 {
		return serviceID[i+1:]
	}
	return serviceID
}

func collectServices(d *xmlDevice, into map[string]Service) {
	for _, s := range d.ServiceList {
		name := shortServiceName(s.ServiceID)
		into[name] = Service{
			Type:       s.ServiceType,
			ID:         s.ServiceID,
			ControlURL: s.ControlURL,
			EventURL:   s.EventSubURL,
			SCPDURL:    s.SCPDURL,
		}
	}
	for i := range d.DeviceList {
		collectServices(&d.DeviceList[i], into)
	}
}

// FetchDevice retrieves and parses http://{ip}:1400/xml/device_description.xml
// into a Device. It returns an error (never a half-populated Device) on any
// network or parse failure.
func FetchDevice(ctx context.Context, ip net.IP) (*Device, error) {
	descURL := fmt.Sprintf("http://%s:%d/xml/device_description.xml", ip.String(), HTTPPort)
	return fetchDeviceURL(ctx, ip, descURL)
}

// FetchDeviceFromLocation parses an SSDP LOCATION URL (which points at the
// description XML) and fetches the device. The IP is taken from the URL.
func FetchDeviceFromLocation(ctx context.Context, location string) (*Device, error) {
	u, err := url.Parse(location)
	if err != nil {
		return nil, fmt.Errorf("parse LOCATION %q: %w", location, err)
	}
	host := u.Hostname()
	ip := net.ParseIP(host)
	if ip == nil {
		return nil, fmt.Errorf("LOCATION host %q is not an IP", host)
	}
	return fetchDeviceURL(ctx, ip, location)
}

func fetchDeviceURL(ctx context.Context, ip net.IP, descURL string) (*Device, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, descURL, nil)
	if err != nil {
		return nil, fmt.Errorf("build request for %s: %w", descURL, err)
	}
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("GET %s: %w", descURL, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GET %s: unexpected status %s", descURL, resp.Status)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", descURL, err)
	}
	return parseDescription(ip, body)
}

func parseDescription(ip net.IP, body []byte) (*Device, error) {
	var root xmlRoot
	if err := xml.Unmarshal(body, &root); err != nil {
		return nil, fmt.Errorf("parse device description: %w", err)
	}
	if root.Device.UDN == "" {
		return nil, fmt.Errorf("device description for %s has no UDN", ip)
	}

	services := make(map[string]Service)
	collectServices(&root.Device, services)

	sw := root.Device.DisplayVersion
	if sw == "" {
		sw = root.Device.SoftwareVersion
	}

	return &Device{
		IP:              ip,
		UDN:             root.Device.UDN,
		FriendlyName:    root.Device.FriendlyName,
		ModelName:       root.Device.ModelName,
		ModelNumber:     root.Device.ModelNumber,
		SoftwareVersion: sw,
		RoomName:        root.Device.RoomName,
		Services:        services,
	}, nil
}

File: packages/core/src/engine/device.ts
// Sonos device-description parsing + S1/S2 generation heuristics.
//
// Ported from backend/internal/sonos/device.go — PURE parsing only. The
// networked FetchDevice (HTTP GET of device_description.xml) lives in a later
// chunk on top of an injected HttpTransport; nothing here touches the network
// or node:*, so this module is part of the RN-facing core surface.
//
// Uses the SHARED fast-xml-parser config from ./soap (parseTagValue:false), so
// every text node — including firmware versions like "15.9" — arrives as a
// Go-faithful string.

import { makeParser } from './soap';
import { textOf, asArray } from './xml';
import type { HttpTransport, Service } from '../sonos';

/**
 * The unencrypted HTTP port every Sonos ZonePlayer exposes for SOAP control,
 * the description XML, GENA event subscriptions and /status/*.
 */
export const HTTPPort = 1400;

/**
 * A discovered Sonos ZonePlayer, populated from its device_description.xml. It
 * is the per-physical-speaker handle; room/group layout comes separately from
 * ZoneGroupTopology.
 */
export interface Device {
  /** Source IP address of the speaker. */
  ip: string;
  /** uuid:RINCON_xxxxxxxxxxxx01400 */
  udn: string;
  friendlyName: string;
  /** e.g. "Sonos PLAY:1" */
  modelName: string;
  /** e.g. "S13" */
  modelNumber: string;
  /** displayVersion (preferred) else softwareVersion, e.g. "15.9" */
  softwareVersion: string;
  /** zone/room name if present in the description */
  roomName: string;
  /**
   * Services indexed by short name (the trailing segment of the serviceId,
   * e.g. "AVTransport", "RenderingControl", "Queue").
   */
  services: Map<string, Service>;
}

/**
 * Returns the bare RINCON UUID (without the "uuid:" prefix) used in x-rincon
 * grouping URIs and topology coordinator references.
 */
export function rincon(device: Device): string {
  return device.udn.startsWith('uuid:') ? device.udn.slice('uuid:'.length) : device.udn;
}

/** http://{ip}:1400 — the root every service path hangs off. */
export function baseURL(device: Device): string {
  return `http://${device.ip}:${HTTPPort}`;
}

/**
 * Reports a best-effort S1/S2 guess. Sonos splits its ecosystem into S1
 * (legacy) and S2 (modern). The dedicated Queue service
 * (urn:schemas-sonos-com:service:Queue:1) is present on S2 players, so we use
 * it as the primary signal and fall back to the firmware major version. The
 * returned string is one of "S2", "S1", or "unknown" — never silently
 * defaulted; callers should show the raw softwareVersion alongside it.
 */
export function generation(device: Device): 'S2' | 'S1' | 'unknown' {
  if (device.services.has('Queue')) {
    return 'S2';
  }
  // Firmware major version: S1 caps out around v11.x; S2 is v12+.
  const maj = firmwareMajor(device.softwareVersion);
  if (maj !== null) {
    return maj >= 12 ? 'S2' : 'S1';
  }
  return 'unknown';
}

/**
 * Parses the leading dotted segment of a firmware version into its major
 * integer. Returns null for empty input or any non-digit in the major segment
 * (mirroring Go's explicit rune scan — no Number() coercion that would accept
 * "11abc" or whitespace).
 */
export function firmwareMajor(version: string): number | null {
  if (version === '') {
    return null;
  }
  const dot = version.indexOf('.');
  const part = dot >= 0 ? version.slice(0, dot) : version;
  if (part === '') {
    return null;
  }
  let n = 0;
  for (const ch of part) {
    if (ch < '0' || ch > '9') {
      return null;
    }
    n = n * 10 + (ch.charCodeAt(0) - 48);
  }
  return n;
}

/**
 * Turns a serviceId like "urn:upnp-org:serviceId:AVTransport" into
 * "AVTransport" (the trailing :segment).
 */
export function shortServiceName(serviceId: string): string {
  const i = serviceId.lastIndexOf(':');
  return i >= 0 ? serviceId.slice(i + 1) : serviceId;
}

// --- description XML parsing ---

function collectServices(device: unknown, into: Map<string, Service>): void {
  if (device === null || typeof device !== 'object') return;
  const node = device as Record<string, unknown>;

  const serviceList = node.serviceList as Record<string, unknown> | undefined;
  if (serviceList) {
    for (const svc of asArray(serviceList.service)) {
      if (svc === null || typeof svc !== 'object') continue;
      const s = svc as Record<string, unknown>;
      const serviceId = textOf(s.serviceId);
      const name = shortServiceName(serviceId);
      into.set(name, {
        serviceType: textOf(s.serviceType),
        serviceId,
        controlURL: textOf(s.controlURL),
        eventSubURL: textOf(s.eventSubURL),
        scpdURL: textOf(s.SCPDURL),
      });
    }
  }

  const deviceList = node.deviceList as Record<string, unknown> | undefined;
  if (deviceList) {
    for (const child of asArray(deviceList.device)) {
      collectServices(child, into);
    }
  }
}

/**
 * Parses a device_description.xml body into a Device. THROWS when the body is
 * unparseable or the root device carries no UDN — never returns a
 * half-populated Device. Prefers displayVersion over softwareVersion for the
 * reported softwareVersion field, matching the Go reference.
 */
export function parseDescription(ip: string, body: string): Device {
  let parsed: Record<string, unknown>;
  try {
    parsed = makeParser().parse(body) as Record<string, unknown>;
  } catch (err) {
    throw new Error(`parse device description: ${(err as Error).message}`);
  }

  const root = parsed.root as Record<string, unknown> | undefined;
  const device = root?.device as Record<string, unknown> | undefined;
  const udn = textOf(device?.UDN);
  if (udn === '') {
    throw new Error(`device description for ${ip} has no UDN`);
  }

  const services = new Map<string, Service>();
  collectServices(device, services);

  const displayVersion = textOf(device?.displayVersion);
  const softwareVersion = displayVersion !== '' ? displayVersion : textOf(device?.softwareVersion);

  return {
    ip,
    udn,
    friendlyName: textOf(device?.friendlyName),
    modelName: textOf(device?.modelName),
    modelNumber: textOf(device?.modelNumber),
    softwareVersion,
    roomName: textOf(device?.roomName),
    services,
  };
}

// --- transport-driven device fetch ---

/**
 * fetchDevice retrieves and parses
 * http://{ip}:1400/xml/device_description.xml via the injected HttpTransport.
 * Ported from Go's FetchDevice. THROWS on a non-200 status or unparseable
 * body — no silent fallback.
 */
export async function fetchDevice(transport: HttpTransport, ip: string): Promise<Device> {
  const descURL = `http://${ip}:${HTTPPort}/xml/device_description.xml`;
  return fetchDeviceURL(transport, ip, descURL);
}

/**
 * fetchDeviceFromLocation parses an SSDP LOCATION URL (which points at the
 * description document), extracts the host IP from it, and fetches it. Ported
 * from Go's FetchDeviceFromLocation. THROWS when the LOCATION host is not an IP
 * literal — no silent fallback.
 */
export async function fetchDeviceFromLocation(
  transport: HttpTransport,
  location: string,
): Promise<Device> {
  const ip = hostFromLocation(location);
  if (ip === null) {
    throw new Error(`LOCATION host in ${location} is not an IP`);
  }
  return fetchDeviceURL(transport, ip, location);
}

async function fetchDeviceURL(
  transport: HttpTransport,
  ip: string,
  descURL: string,
): Promise<Device> {
  const resp = await transport.request({ method: 'GET', url: descURL });
  if (resp.status !== 200) {
    throw new Error(`GET ${descURL}: unexpected status ${resp.status}`);
  }
  return parseDescription(ip, resp.body);
}

/**
 * Extracts a literal-IP host from an `http://{host}[:port]/...` URL using the
 * global URL parser (available in RN/browser/node — no node:* import). Returns
 * the host only when it parses as an IPv4 or bracketed-IPv6 literal; otherwise
 * null (a hostname is rejected, mirroring Go's net.ParseIP gate).
 */
function hostFromLocation(location: string): string | null {
  let host: string;
  try {
    host = new URL(location).hostname;
  } catch {
    return null;
  }
  // URL strips the brackets off an IPv6 host; re-detect both families.
  if (/^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.test(host)) {
    return host.split('.').every((o) => Number(o) <= 255) ? host : null;
  }
  if (host.includes(':') && /^[0-9a-fA-F:.]+$/.test(host)) {
    return host;
  }
  return null;
}
